docker/gcp_pudl_etl.sh

```diff
@@ -97,7 +97,7 @@ if [[ ${PIPESTATUS[0]} == 0 ]]; then
     # Deploy the updated data to datasette
     if [ $GITHUB_REF = "dev" ]; then
         gcloud config set run/region us-central1
-        source ~/devtools/datasette/publish.sh
+        python ~/devtools/datasette/publish.py
```

Comment:
How can we be notified if this script fails so we don't run into the scenario where we're unaware of many datasette deployment failures?

Reply:
Ah, since the `notify_slack` happens before we even try to publish - yeah, that's a good point, though I'm not sure if we had this functionality in the old world.

Do you think the following diff would work?

```diff
diff --git a/docker/gcp_pudl_etl.sh b/docker/gcp_pudl_etl.sh
index 5e0929b33..0fe5c26fe 100644
--- a/docker/gcp_pudl_etl.sh
+++ b/docker/gcp_pudl_etl.sh
@@ -85,10 +85,8 @@ function notify_slack() {
 # 2>&1 redirects stderr to stdout.
 run_pudl_etl 2>&1 | tee $LOGFILE
 
-# Notify slack if the etl succeeded.
+# if pipeline is successful, distribute + publish datasette
 if [[ ${PIPESTATUS[0]} == 0 ]]; then
-    notify_slack "success"
-
     # Dump outputs to s3 bucket if branch is dev or build was triggered by a tag
     if [ $GITHUB_ACTION_TRIGGER = "push" ] || [ $GITHUB_REF = "dev" ]; then
         copy_outputs_to_distribution_bucket
@@ -99,8 +97,13 @@ if [[ ${PIPESTATUS[0]} == 0 ]]; then
         gcloud config set run/region us-central1
         python ~/devtools/datasette/publish.py
     fi
-else
-    notify_slack "failure"
 fi
 
+# Notify slack about entire pipeline's success or failure;
+# PIPESTATUS[0] either refers to the failed ETL run or the last distribution
+# task that was run above
+if [[ ${PIPESTATUS[0]} == 0 ]]; then notify_slack "success" else notify_slack
+    "failure" fi
+
+
 shutdown_vm
```

Annoyingly, the most straightforward way I can imagine testing it is "include `assert False` at the end of `publish.py` and wait for the nightly build to fail there" - any ideas?

Reply:
To test it out on this branch, you could run the fast ETL, skip the tests and include assert False at the end of `publish.py`. 

That logic looks correct! Any suggestions for how to capture the logs for the distribute + publish datasette steps? With your current proposal if, the ETL will succeed and the datasette deployment fails, we'll get a notification that something failed but won't find any failures in the logs.

Reply:
I captured the logs by `tee -a $LOGFILE` so that the publish logs just end up at the end of the `pudl-etl.log`. This worked for debugging a failed deploy just now so I think that's good enough :)

Reply:
Love it thank you!